passes/techmap/extract_reduce.cc

```diff
@@ -154,36 +180,54 @@ struct ExtractReducePass : public Pass
 					cur_supercell.insert(x);
 
 					auto a = sigmap(x->getPort("\\A"));
-					log_assert(a.size() == 1);
-					// Must have only one sink
-					// XXX: Check that it is indeed this node?
-					if (sig_to_sink[a[0]].size() + port_sigs.count(a[0]) == 1)
+					if (a.size() == 1)
 					{
-						Cell* cell_a = sig_to_driver[a[0]];
-						if (((cell_a->type == "$_AND_" && gt == GateType::And) ||
-							(cell_a->type == "$_OR_" && gt == GateType::Or) ||
-							(cell_a->type == "$_XOR_" && gt == GateType::Xor)))
-						{
-							// The cell here is the correct type, and it's definitely driving only
-							// this current cell.
-							bfs_queue.push_back(cell_a);
+						if (a[0].wire != NULL){
+							log_assert(a.size() == 1);
+							// Must have only one sink
+							// XXX: Check that it is indeed this node?
+							if (sig_to_sink[a[0]].size() + port_sigs.count(a[0]) == 1)
+							{
+								Cell* cell_a = sig_to_driver[a[0]];
+								if (((cell_a->type == "$_AND_" && gt == GateType::And) ||
+									 (cell_a->type == "$_OR_" && gt == GateType::Or) ||
+									 (cell_a->type == "$_XOR_" && gt == GateType::Xor) ||
+									 (cell_a->type == "$and" && gt == GateType::And) ||
+									 (cell_a->type == "$or" && gt == GateType::Or) ||
+									 (cell_a->type == "$xor" && gt == GateType::Xor)
```

Comment:
I think you should move a copy of the width checking from line 148 to here instead.

Reply:
Refactored.